File: test/p11rand-test.c
/* Tiny CLI driver: dlopen() the module, list slots, ask the first slot's
 * token for 32 random bytes, dump them as hex.
 *
 * Usage: p11rand-test ./libp11rand.so [byte_count]
 */

#include "../include/pkcs11.h"

#include <dlfcn.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void die(const char *what, CK_RV rv)
{
    fprintf(stderr, "%s failed: 0x%lx\n", what, (unsigned long)rv);
    exit(2);
}

static void show_field(const char *label, const unsigned char *p, size_t n)
{
    /* PKCS#11 strings are space-padded; trim trailing spaces. */
    while (n && p[n - 1] == ' ') n--;
    printf("  %-12s %.*s\n", label, (int)n, p);
}

int main(int argc, char **argv)
{
    if (argc < 2) {
        fprintf(stderr, "usage: %s <module.so> [nbytes]\n", argv[0]);
        return 1;
    }
    size_t n = (argc >= 3) ? strtoul(argv[2], NULL, 0) : 32;

    void *h = dlopen(argv[1], RTLD_NOW);
    if (!h) { fprintf(stderr, "dlopen: %s\n", dlerror()); return 2; }

    CK_RV (*GetFunctionList)(CK_FUNCTION_LIST_PTR_PTR);
    *(void **)&GetFunctionList = dlsym(h, "C_GetFunctionList");
    if (!GetFunctionList) { fprintf(stderr, "dlsym C_GetFunctionList: %s\n", dlerror()); return 2; }

    CK_FUNCTION_LIST_PTR f = NULL;
    CK_RV rv = GetFunctionList(&f);
    if (rv != CKR_OK) die("C_GetFunctionList", rv);

    rv = f->C_Initialize(NULL);
    if (rv != CKR_OK) die("C_Initialize", rv);

    /* Enumerate every reader (tokenPresent=FALSE) so we can show empty
     * slots; then enumerate again with tokenPresent=TRUE to pick a slot
     * whose card actually answers GET CHALLENGE. */
    CK_ULONG nall = 0;
    rv = f->C_GetSlotList(CK_FALSE, NULL, &nall);
    if (rv != CKR_OK) die("C_GetSlotList(size, all)", rv);
    CK_SLOT_ID *all = nall ? calloc(nall, sizeof(*all)) : NULL;
    if (nall) {
        rv = f->C_GetSlotList(CK_FALSE, all, &nall);
        if (rv != CKR_OK) die("C_GetSlotList(all)", rv);
    }

    printf("readers visible: %lu\n", (unsigned long)nall);
    for (CK_ULONG i = 0; i < nall; i++) {
        CK_SLOT_INFO si;
        if (f->C_GetSlotInfo(all[i], &si) != CKR_OK) continue;
        const char *mark = (si.flags & CKF_TOKEN_PRESENT) ? "[token]" : "[empty]";
        printf("[%lu] %s flags=0x%lx\n", (unsigned long)all[i], mark,
               (unsigned long)si.flags);
        show_field("reader", si.slotDescription, sizeof(si.slotDescription));
    }

    CK_ULONG nuse = 0;
    rv = f->C_GetSlotList(CK_TRUE, NULL, &nuse);
    if (rv != CKR_OK) die("C_GetSlotList(size, present)", rv);
    if (nuse == 0) {
        fprintf(stderr, "no usable token in any reader\n");
        free(all);
        f->C_Finalize(NULL);
        return 3;
    }
    CK_SLOT_ID *use = calloc(nuse, sizeof(*use));
    if (!use) return 2;
    rv = f->C_GetSlotList(CK_TRUE, use, &nuse);
    if (rv != CKR_OK) die("C_GetSlotList(present)", rv);

    CK_SLOT_ID picked = use[0];
    printf("using slot %lu (first with a usable token, of %lu)\n",
           (unsigned long)picked, (unsigned long)nuse);

    CK_SESSION_HANDLE sess = 0;
    rv = f->C_OpenSession(picked, CKF_SERIAL_SESSION, NULL, NULL, &sess);
    if (rv != CKR_OK) die("C_OpenSession", rv);

    CK_TOKEN_INFO ti;
    if (f->C_GetTokenInfo(picked, &ti) == CKR_OK) {
        puts("token:");
        show_field("label", ti.label,          sizeof(ti.label));
        show_field("model", ti.model,          sizeof(ti.model));
        show_field("mfr",   ti.manufacturerID, sizeof(ti.manufacturerID));
        printf("  flags        0x%lx\n", (unsigned long)ti.flags);
    }

    uint8_t *buf = calloc(1, n);
    if (!buf) return 2;
    rv = f->C_GenerateRandom(sess, buf, (CK_ULONG)n);
    if (rv != CKR_OK) die("C_GenerateRandom", rv);

    printf("random[%zu]: ", n);
    for (size_t i = 0; i < n; i++) printf("%02x", buf[i]);
    putchar('\n');

    free(buf);
    free(use);
    free(all);
    f->C_CloseSession(sess);
    f->C_Finalize(NULL);
    dlclose(h);
    return 0;
}

File: crates/p11rand-tools/src/bin/p11rand-test.rs
//! Tiny CLI driver: dlopen the module, list slots, ask the first slot's
//! token for 32 random bytes, dump them as hex.
//!
//! Usage: `p11rand-test ./libp11rand.so [byte_count]`

#![allow(non_snake_case)]

use std::process::ExitCode;

use p11rand_tools::cryptoki_sys::{self, *};
use p11rand_tools::{load_module, trim_pkcs11};

fn die(what: &str, rv: CK_RV) -> ! {
    eprintln!("{what} failed: 0x{rv:x}");
    std::process::exit(2);
}

fn show_field(label: &str, field: &[u8]) {
    println!("  {label:<12} {}", trim_pkcs11(field));
}

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("usage: {} <module.so> [nbytes]", args[0]);
        return ExitCode::from(1);
    }
    let n: usize = if args.len() >= 3 {
        args[2].parse().unwrap_or(32)
    } else {
        32
    };

    let m = match load_module(&args[1]) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("{e}");
            return ExitCode::from(2);
        }
    };
    let f = m.funcs;
    unsafe {
        let rv = f.C_Initialize.unwrap()(std::ptr::null_mut());
        if rv != CKR_OK {
            die("C_Initialize", rv);
        }

        // Enumerate every reader (tokenPresent=false) so we can show empty
        // slots; then enumerate again with tokenPresent=true to pick a slot
        // whose token is actually usable.
        let mut nall: CK_ULONG = 0;
        let rv = f.C_GetSlotList.unwrap()(CK_FALSE, std::ptr::null_mut(), &mut nall);
        if rv != CKR_OK {
            die("C_GetSlotList(size, all)", rv);
        }
        let mut all: Vec<CK_SLOT_ID> = vec![0; nall as usize];
        if nall > 0 {
            let rv = f.C_GetSlotList.unwrap()(CK_FALSE, all.as_mut_ptr(), &mut nall);
            if rv != CKR_OK {
                die("C_GetSlotList(all)", rv);
            }
            all.truncate(nall as usize);
        }

        println!("readers visible: {}", nall);
        for &slot in &all {
            let mut si: CK_SLOT_INFO = std::mem::zeroed();
            if f.C_GetSlotInfo.unwrap()(slot, &mut si) != CKR_OK {
                continue;
            }
            let mark = if (si.flags & cryptoki_sys::CKF_TOKEN_PRESENT) != 0 {
                "[token]"
            } else {
                "[empty]"
            };
            println!("[{slot}] {mark} flags=0x{:x}", si.flags);
            show_field("reader", &si.slotDescription);
        }

        let mut nuse: CK_ULONG = 0;
        let rv = f.C_GetSlotList.unwrap()(CK_TRUE, std::ptr::null_mut(), &mut nuse);
        if rv != CKR_OK {
            die("C_GetSlotList(size, present)", rv);
        }
        if nuse == 0 {
            eprintln!("no usable token in any reader");
            f.C_Finalize.unwrap()(std::ptr::null_mut());
            return ExitCode::from(3);
        }
        let mut use_slots: Vec<CK_SLOT_ID> = vec![0; nuse as usize];
        let rv = f.C_GetSlotList.unwrap()(CK_TRUE, use_slots.as_mut_ptr(), &mut nuse);
        if rv != CKR_OK {
            die("C_GetSlotList(present)", rv);
        }

        let picked = use_slots[0];
        println!(
            "using slot {} (first with a usable token, of {})",
            picked, nuse
        );

        let mut sess: CK_SESSION_HANDLE = 0;
        let rv = f.C_OpenSession.unwrap()(
            picked,
            CKF_SERIAL_SESSION,
            std::ptr::null_mut(),
            None,
            &mut sess,
        );
        if rv != CKR_OK {
            die("C_OpenSession", rv);
        }

        let mut ti: CK_TOKEN_INFO = std::mem::zeroed();
        if f.C_GetTokenInfo.unwrap()(picked, &mut ti) == CKR_OK {
            println!("token:");
            show_field("label", &ti.label);
            show_field("model", &ti.model);
            show_field("mfr", &ti.manufacturerID);
            println!("  flags        0x{:x}", ti.flags);
        }

        let mut buf = vec![0u8; n];
        let rv = f.C_GenerateRandom.unwrap()(sess, buf.as_mut_ptr(), n as CK_ULONG);
        if rv != CKR_OK {
            die("C_GenerateRandom", rv);
        }

        print!("random[{n}]: ");
        for b in &buf {
            print!("{b:02x}");
        }
        println!();

        f.C_CloseSession.unwrap()(sess);
        f.C_Finalize.unwrap()(std::ptr::null_mut());
    }
    ExitCode::SUCCESS
}
